In [6]:
# First let's Load the csv's
import pandas as pd

receipts = pd.read_csv ('C:/Users/Parth Parsana/OneDrive/Desktop/Interview prep/Fetch reward/receipts.csv')
receipts

_id  bonusPointsEarned  \
0     5ff1e1eb0a720f0523000575              500.0   
1     5ff1e1bb0a720f052300056b              150.0   
2     5ff1e1f10a720f052300057a                5.0   
3     5ff1e1ee0a7214ada100056f                5.0   
4     5ff1e1d20a7214ada1000561                5.0   
...                        ...                ...   
1114  603cc0630a720fde100003e6               25.0   
1115  603d0b710a720fde1000042a                NaN   
1116  603cf5290a720fde10000413                NaN   
1117  603ce7100a7217c72c000405               25.0   
1118  603c4fea0a7217c72c000389                NaN   

                                bonusPointsEarnedReason  \
0     Receipt number 2 completed, bonus point schedu...   
1     Receipt number 5 completed, bonus point schedu...   
2                            All-receipts receipt bonus   
3                            All-receipts receipt bonus   
4                            All-receipts receipt bonus   
...                                                 ...   
1114                        COMPLETE_NONPARTNER_RECEIPT   
1115                                                NaN   
1116                                                NaN   
1117                        COMPLETE_NONPARTNER_RECEIPT   
1118                                                NaN   

                         createDate                    dateScanned  \
0     2021-01-03 15:25:31.000000000  2021-01-03 15:25:31.000000000   
1     2021-01-03 15:24:43.000000000  2021-01-03 15:24:43.000000000   
2     2021-01-03 15:25:37.000000000  2021-01-03 15:25:37.000000000   
3     2021-01-03 15:25:34.000000000  2021-01-03 15:25:34.000000000   
4     2021-01-03 15:25:06.000000000  2021-01-03 15:25:06.000000000   
...                             ...                            ...   
1114  2021-03-01 10:22:27.000000000  2021-03-01 10:22:27.000000000   
1115  2021-03-01 15:42:41.873000000  2021-03-01 15:42:41.873000000   
1116  2021-03-01 14:07:37.664000000  2021-03-01 14:07:37.664000000   
1117  2021-03-01 13:07:28.000000000  2021-03-01 13:07:28.000000000   
1118  2021-03-01 02:22:34.962000000  2021-03-01 02:22:34.962000000   

                       finishedDate                     modifyDate  \
0     2021-01-03 15:25:31.000000000  2021-01-03 15:25:36.000000000   
1     2021-01-03 15:24:43.000000000  2021-01-03 15:24:48.000000000   
2                               NaN  2021-01-03 15:25:42.000000000   
3     2021-01-03 15:25:34.000000000  2021-01-03 15:25:39.000000000   
4     2021-01-03 15:25:11.000000000  2021-01-03 15:25:11.000000000   
...                             ...                            ...   
1114                            NaN  2021-03-01 10:22:28.000000000   
1115                            NaN  2021-03-01 15:42:41.873000000   
1116                            NaN  2021-03-01 14:07:37.664000000   
1117                            NaN  2021-03-01 13:07:29.000000000   
1118                            NaN  2021-03-01 02:22:34.962000000   

                  pointsAwardedDate  pointsEarned         purchaseDate  \
0     2021-01-03 15:25:31.000000000         500.0  2021-01-03 00:00:00   
1     2021-01-03 15:24:43.000000000         150.0  2021-01-02 15:24:43   
2                               NaN           5.0  2021-01-03 00:00:00   
3     2021-01-03 15:25:34.000000000           5.0  2021-01-03 00:00:00   
4     2021-01-03 15:25:06.000000000           5.0  2021-01-02 15:25:06   
...                             ...           ...                  ...   
1114                            NaN          25.0  2020-08-17 00:00:00   
1115                            NaN           NaN                  NaN   
1116                            NaN           NaN                  NaN   
1117                            NaN          25.0  2020-08-17 00:00:00   
1118                            NaN           NaN                  NaN   

      purchasedItemCount rewardsReceiptStatus  totalSpent  \
0                    5.0             FINISHED    

## Looking at this table, we can see many datetime formats are present like createdate, finishedDate, modifyDate, datescanned, purchasedate, pointsAwardedDate. We can remove createDate because it is same as dateScanned. And ideally the day event is created is going to be when new receipt is scanned.

In [7]:
users = pd.read_csv ('C:/Users/Parth Parsana/OneDrive/Desktop/Interview prep/Fetch reward/users.csv')
users

_id  active              createdDate  \
0    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04.800   
1    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04.800   
2    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04.800   
3    5ff1e1eacfcf6c399c274ae6    True  2021-01-03 15:25:30.554   
4    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04.800   
..                        ...     ...                      ...   
490  54943462e4b07e684157a532    True  2014-12-19 14:21:22.381   
491  54943462e4b07e684157a532    True  2014-12-19 14:21:22.381   
492  54943462e4b07e684157a532    True  2014-12-19 14:21:22.381   
493  54943462e4b07e684157a532    True  2014-12-19 14:21:22.381   
494  54943462e4b07e684157a532    True  2014-12-19 14:21:22.381   

                         lastLogin         role signUpSource state  
0    2021-01-03 15:25:37.857999872     consumer        Email    WI  
1    2021-01-03 15:25:37.857999872     consumer        Email    WI  
2    2021-01-03 15:25:37.857999872     consumer        Email    WI  
3    2021-01-03 15:25:30.596999936     consumer        Email    WI  
4    2021-01-03 15:25:37.857999872     consumer        Email    WI  
..                             ...          ...          ...   ...  
490  2021-03-05 16:52:23.204000000  fetch-staff          NaN   NaN  
491  2021-03-05 16:52:23.204000000  fetch-staff          NaN   NaN  
492  2021-03-05 16:52:23.204000000  fetch-staff          NaN   NaN  
493  2021-03-05 16:52:23.204000000  fetch-staff          NaN   NaN  
494  2021-03-05 16:52:23.204000000  fetch-staff          NaN   NaN  

[495 rows x 7 columns]

In [8]:
brands = pd.read_csv ('C:/Users/Parth Parsana/OneDrive/Desktop/Interview prep/Fetch reward/brands.csv')
brands

_id       barcode            category  \
0     601ac115be37ce2ead437551  511111019862              Baking   
1     601c5460be37ce2ead43755f  511111519928           Beverages   
2     601ac142be37ce2ead43755d  511111819905              Baking   
3     601ac142be37ce2ead43755a  511111519874              Baking   
4     601ac142be37ce2ead43755e  511111319917      Candy & Sweets   
...                        ...           ...                 ...   
1162  5f77274dbe37ce6b592e90c0  511111116752              Baking   
1163  5dc1fca91dda2c0ad7da64ae  511111706328  Breakfast & Cereal   
1164  5f494c6e04db711dd8fe87e7  511111416173      Candy & Sweets   
1165  5a021611e4b00efe02b02a57  511111400608             Grocery   
1166  6026d757be37ce6369301468  511111019930              Baking   

          categoryCode                       name  topBrand  \
0               BAKING  test brand @1612366101024     False   
1            BEVERAGES                  Starbucks     False   
2               BAKING  test brand @1612366146176     False   
3               BAKING  test brand @1612366146051     False   
4     CANDY_AND_SWEETS  test brand @1612366146827     False   
...                ...                        ...       ...   
1162            BAKING  test brand @1601644365844     False   
1163               NaN        Dippin Dots® Cereal     False   
1164  CANDY_AND_SWEETS  test brand @1598639215217     False   
1165               NaN          LIPTON TEA Leaves     False   
1166            BAKING  test brand @1613158231643     False   

                          brandCode                    cpg_Id cpf_Ref  
0                               NaN  601ac114be37ce2ead437550    Cogs  
1                         STARBUCKS  5332f5fbe4b03c9a25efd0ba    Cogs  
2     TEST BRANDCODE @1612366146176  601ac142be37ce2ead437559    Cogs  
3     TEST BRANDCODE @1612366146051  601ac142be37ce2ead437559    Cogs  
4     TEST BRANDCODE @1612366146827  5332fa12e4b03c9a25efd1e7    Cogs  
...                             ...                       ...     ...  
1162                            NaN  5f77274dbe37ce6b592e90bf    Cogs  
1163             DIPPIN DOTS CEREAL  53e10d6368abd3c7065097cc    Cogs  
1164  TEST BRANDCODE @1598639215217  5332fa12e4b03c9a25efd1e7    Cogs  
1165              LIPTON TEA Leaves  5332f5f6e4b03c9a25efd0b4    Cogs  
1166  TEST BRANDCODE @1613158231644  6026d757be37ce6369301467    Cogs  

[1167 rows x 9 columns]

In [9]:
items = pd.read_csv ('C:/Users/Parth Parsana/OneDrive/Desktop/Interview prep/Fetch reward/items.csv')
items

userId              dateScanned  \
0     5ff1e1eacfcf6c399c274ae6      2021-01-03 15:25:31   
1     5ff1e194b6a9d73a3a9f1052      2021-01-03 15:24:43   
2     5ff1e194b6a9d73a3a9f1052      2021-01-03 15:24:43   
3     5ff1e1f1cfcf6c399c274b0b      2021-01-03 15:25:37   
4     5ff1e1eacfcf6c399c274ae6      2021-01-03 15:25:34   
...                        ...                      ...   
6936  5fc961c3b8cfca11a077dd33  2021-03-01 10:32:28.000   
6937  5fc961c3b8cfca11a077dd33  2021-03-01 10:22:27.000   
6938  5fc961c3b8cfca11a077dd33  2021-03-01 10:22:27.000   
6939  5fc961c3b8cfca11a077dd33  2021-03-01 13:07:28.000   
6940  5fc961c3b8cfca11a077dd33  2021-03-01 13:07:28.000   

                           _id  level_1       barcode  \
0     5ff1e1eb0a720f0523000575        0          4011   
1     5ff1e1bb0a720f052300056b        0          4011   
2     5ff1e1bb0a720f052300056b        1  028400642255   
3     5ff1e1f10a720f052300057a        0           NaN   
4     5ff1e1ee0a7214ada100056f        0          4011   
...                        ...      ...           ...   
6936  603cc2bc0a720fde100003e9        1    B07BRRLSVC   
6937  603cc0630a720fde100003e6        0    B076FJ92M4   
6938  603cc0630a720fde100003e6        1    B07BRRLSVC   
6939  603ce7100a7217c72c000405        0    B076FJ92M4   
6940  603ce7100a7217c72c000405        1    B07BRRLSVC   

                                            description  finalPrice  \
0                                        ITEM NOT FOUND       26.00   
1                                        ITEM NOT FOUND        1.00   
2     DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...       10.00   
3                                                   NaN         NaN   
4                                        ITEM NOT FOUND       28.00   
...                                                 ...         ...   
6936  thindust summer face mask - sun protection nec...       11.99   
6937  mueller austria hypergrind precision electric ...       22.97   
6938  thindust summer face mask - sun protection nec...       11.99   
6939  mueller austria hypergrind precision electric ...       22.97   
6940  thindust summer face mask - sun protection nec...       11.99   

      itemPrice needsFetchReview  partnerItemId  ... itemNumber  \
0         26.00            False              1  ...        NaN   
1          1.00              NaN              1  ...        NaN   
2         10.00             True              2  ...        NaN   
3           NaN            False              1  ...        NaN   
4         28.00            False              1  ...        NaN   
...         ...              ...            ...  ...        ...   
6936      11.99              NaN              1  ...        NaN   
6937      22.97              NaN              0  ...        NaN   
6938      11.99              NaN              1  ...        NaN   
6939      22.97              NaN              0  ...        NaN   
6940      11.99              NaN              1  ...        NaN   

      originalMetaBriteQuantityPurchased  pointsEarned targetPrice  \
0                                    NaN           NaN         NaN   
1                                    NaN           NaN         NaN   
2                                    NaN           NaN         NaN   
3                                    NaN           NaN         NaN   
4                                    NaN           NaN         NaN   
...                                  ...           ...         ...   
6936                                 NaN           NaN         NaN   
6937                                 NaN           NaN         NaN   
6938                                 NaN           NaN         NaN   
6939                                 NaN           NaN         NaN   
6940                                 NaN           NaN         NaN   

      competitiveProduct  originalFinalPrice originalMetaBriteItemPrice  \
0                    NaN                 NaN            

## Just looking at the items table we can see a major issues with data quality like:
##  Items table which is derived from receipts table has incomplete barcode. Since there are two types of barcode:UPC-E and UPC-A. The UPC-E can encode only six digits, while UPC-A encodes twelve digits. Neither of those conditions are fulfilled for many items


In [20]:
len(users['_id'].unique())

212

In [15]:
print(receipts.shape)
receipts.isnull().sum()

(1119, 14)


_id                          0
bonusPointsEarned          575
bonusPointsEarnedReason    575
createDate                   0
dateScanned                  0
finishedDate               551
modifyDate                   0
pointsAwardedDate          582
pointsEarned               510
purchaseDate               448
purchasedItemCount         484
rewardsReceiptStatus         0
totalSpent                 435
userId                       0
dtype: int64

In [16]:
print(users.shape)
users.isnull().sum()

(495, 7)


_id              0
active           0
createdDate      0
lastLogin       62
role             0
signUpSource    48
state           56
dtype: int64

In [23]:
len(users['_id'].unique())

212

### The primary key used as userId (_id) in user table has only 212 unique values that means the rest 283 _ids are duplicates. This can have negative impact if we use this primary key for linking other tables.

In [17]:
print(brands.shape)
brands.isnull().sum()

(1167, 9)


_id               0
barcode           0
category        155
categoryCode    650
name              0
topBrand          0
brandCode       269
cpg_Id            0
cpf_Ref           0
dtype: int64

In [14]:
print(items.shape)
items.isnull().sum()

(6941, 38)


userId                                   0
dateScanned                              0
_id                                      0
level_1                                  0
barcode                               3851
description                            381
finalPrice                             174
itemPrice                              174
needsFetchReview                      6128
partnerItemId                            0
preventTargetGapPoints                6583
quantityPurchased                      174
userFlaggedBarcode                    6604
userFlaggedNewItem                    6618
userFlaggedPrice                      6642
userFlaggedQuantity                   6642
needsFetchReviewReason                6722
pointsNotAwardedReason                6601
pointsPayerId                         5674
rewardsGroup                          5210
rewardsProductPartnerId               4672
userFlaggedDescription                6787
originalMetaBriteBarcode              6917
originalMet

### From the total null values we can see that all the table have some null values like in:
### Receipts: bonusPointsEarned, purchasedItemCount and totalSpent have a lot of null values which are very significant for the questions asked.
### Users: signupScore, state and lastlogin are the only fields with null value
### Brands: brandCode, categoryCode and category are the only fields with null value. 
### Items: Many fields have null values. But most imporatant is BarCode has many null values whichi is used to link the items table so has significant impact on the answers.

### Also all time format are given in unix-millisecond format which is quicker to process than datetime format but is not so easily readble. So, we have to convert the format before we perform any analysis

### Additionally, the rewardsReceiptItemList which is removed from receipts has a list of items present inside which is a very computationally heavy way of processing this data directly. That is why I have put that list as a new table Items and linked it with receipts table using  the receiptId (_id)